File: src/auth/oauth.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal
// This program is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version. See <https://www.gnu.org/licenses/>.

/**
 * OAuth 2.0 / OpenID Connect login for Google, Microsoft, and LinkedIn.
 *
 * Authorization-code flow, no extra OAuth SDK:
 *   GET  /auth/providers           → which providers are configured (for the UI)
 *   GET  /auth/:provider/login     → redirect to the provider with a signed state
 *   GET  /auth/:provider/callback  → exchange code, fetch userinfo, map to a
 *                                    LawyerProfile (by email), set a signed
 *                                    session cookie, redirect back to the UI
 *   POST /auth/logout              → clear the session cookie
 *
 * The session is a signed, httpOnly cookie holding the SessionUser — stateless,
 * no server-side session store. Profiles are auto-provisioned on first login
 * (partner if the email is in ADMIN_EMAILS, otherwise lawyer).
 *
 * NOTE: this is wired and type-checked but cannot be live-tested without OAuth
 * app credentials registered with each provider (client id/secret + redirect
 * URI). See the README "Deploying with login" section.
 */

import { randomUUID } from "crypto";
import type { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";
import { Config } from "../config.js";
import { logger } from "../logger.js";
import type { Orchestrator } from "../orchestrator.js";
import type { SessionUser } from "../types.js";

const SESSION_COOKIE = "bm_session";
const STATE_COOKIE = "bm_oauth_state";

type ProviderKey = "google" | "microsoft" | "linkedin";

interface ProviderSpec {
  authUrl: string;
  tokenUrl: string;
  userInfoUrl: string;
  scope: string;
  /** Map the provider's userinfo payload to a normalized identity. */
  mapUser: (info: Record<string, unknown>) => { sub: string; email: string; name: string };
}

const PROVIDERS: Record<ProviderKey, ProviderSpec> = {
  google: {
    authUrl: "https://accounts.google.com/o/oauth2/v2/auth",
    tokenUrl: "https://oauth2.googleapis.com/token",
    userInfoUrl: "https://openidconnect.googleapis.com/v1/userinfo",
    scope: "openid email profile",
    mapUser: (i) => ({ sub: String(i.sub), email: String(i.email ?? ""), name: String(i.name ?? i.email ?? "User") }),
  },
  microsoft: {
    authUrl: "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
    tokenUrl: "https://login.microsoftonline.com/common/oauth2/v2.0/token",
    userInfoUrl: "https://graph.microsoft.com/oidc/userinfo",
    scope: "openid email profile",
    mapUser: (i) => ({ sub: String(i.sub), email: String(i.email ?? i.preferred_username ?? ""), name: String(i.name ?? i.email ?? "User") }),
  },
  linkedin: {
    authUrl: "https://www.linkedin.com/oauth/v2/authorization",
    tokenUrl: "https://www.linkedin.com/oauth/v2/accessToken",
    userInfoUrl: "https://api.linkedin.com/v2/userinfo",
    scope: "openid email profile",
    mapUser: (i) => ({ sub: String(i.sub), email: String(i.email ?? ""), name: String(i.name ?? i.email ?? "User") }),
  },
};

const creds = (p: ProviderKey) => Config.auth.providers[p];
const isConfigured = (p: ProviderKey): boolean => !!creds(p).clientId && !!creds(p).clientSecret;
const redirectUri = (p: ProviderKey) => `${Config.auth.baseUrl.replace(/\/$/, "")}/auth/${p}/callback`;

const cookieOpts = (maxAgeSec: number) => ({
  signed: true, httpOnly: true as const, sameSite: "lax" as const,
  secure: Config.auth.baseUrl.startsWith("https"), path: "/", maxAge: maxAgeSec,
});

/** Read the signed session cookie → SessionUser, or null. */
export function readSessionCookie(req: FastifyRequest): SessionUser | null {
  const raw = (req.cookies as Record<string, string> | undefined)?.[SESSION_COOKIE];
  if (!raw) return null;
  const unsigned = (req as unknown as { unsignCookie: (v: string) => { valid: boolean; value: string | null } }).unsignCookie(raw);
  if (!unsigned.valid || !unsigned.value) return null;
  try { return JSON.parse(unsigned.value) as SessionUser; } catch { return null; }
}

export function registerAuthRoutes(app: FastifyInstance, orchestrator: Orchestrator): void {
  app.get("/auth/providers", async () => ({
    google: isConfigured("google"),
    microsoft: isConfigured("microsoft"),
    linkedin: isConfigured("linkedin"),
  }));

  app.get("/auth/:provider/login", async (req, reply) => {
    const provider = (req.params as { provider: string }).provider as ProviderKey;
    const spec = PROVIDERS[provider];
    if (!spec || !isConfigured(provider)) return reply.code(404).send({ error: "Provider not configured" });

    const state = randomUUID();
    reply.setCookie(STATE_COOKIE, state, cookieOpts(600));
    const url = new URL(spec.authUrl);
    url.searchParams.set("response_type", "code");
    url.searchParams.set("client_id", creds(provider).clientId);
    url.searchParams.set("redirect_uri", redirectUri(provider));
    url.searchParams.set("scope", spec.scope);
    url.searchParams.set("state", state);
    return reply.redirect(url.toString());
  });

  app.get("/auth/:provider/callback", async (req, reply) => {
    const provider = (req.params as { provider: string }).provider as ProviderKey;
    const spec = PROVIDERS[provider];
    if (!spec || !isConfigured(provider)) return reply.code(404).send({ error: "Provider not configured" });

    const { code, state } = req.query as { code?: string; state?: string };
    const expected = readSignedCookie(req, STATE_COOKIE);
    if (!code || !state || state !== expected) return reply.code(400).send({ error: "Invalid OAuth state" });
    reply.clearCookie(STATE_COOKIE, { path: "/" });

    try {
      // Exchange the code for an access token.
      const tokenRes = await fetch(spec.tokenUrl, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded", Accept: "application/json" },
        body: new URLSearchParams({
          grant_type: "authorization_code",
          code,
          redirect_uri: redirectUri(provider),
          client_id: creds(provider).clientId,
          client_secret: creds(provider).clientSecret,
        }),
      });
      if (!tokenRes.ok) throw new Error(`token exchange failed (${tokenRes.status})`);
      const token = (await tokenRes.json()) as { access_token?: string };
      if (!token.access_token) throw new Error("no access_token in response");

      // Fetch the user's profile.
      const infoRes = await fetch(spec.userInfoUrl, { headers: { Authorization: `Bearer ${token.access_token}` } });
      if (!infoRes.ok) throw new Error(`userinfo failed (${infoRes.status})`);
      const identity = spec.mapUser((await infoRes.json()) as Record<string, unknown>);
      if (!identity.email) throw new Error("provider returned no email");

      // Map to a lawyer profile (auto-provision on first login).
      let profile = orchestrator.profiles.getByEmail(identity.email);
      if (!profile) {
        const role = Config.auth.adminEmails.includes(identity.email.toLowerCase()) ? "partner" : "lawyer";
        profile = await orchestrator.profiles.create({ name: identity.name, email: identity.email, role });
      }

      const user: SessionUser = { profileId: profile.id, name: profile.name, email: profile.email, role: profile.role };
      reply.setCookie(SESSION_COOKIE, JSON.stringify(user), cookieOpts(60 * 60 * 12));
      logger.info("OAuth login", { provider, email: identity.email, role: profile.role });
      return reply.redirect(Config.auth.uiUrl);
    } catch (err) {
      logger.warn("OAuth callback failed", { provider, error: (err as Error).message });
      return reply.redirect(`${Config.auth.uiUrl}?auth_error=1`);
    }
  });

  app.post("/auth/logout", async (_req, reply) => {
    reply.clearCookie(SESSION_COOKIE, { path: "/" });
    return { ok: true };
  });
}

function readSignedCookie(req: FastifyRequest, name: string): string | null {
  const raw = (req.cookies as Record<string, string> | undefined)?.[name];
  if (!raw) return null;
  const unsigned = (req as unknown as { unsignCookie: (v: string) => { valid: boolean; value: string | null } }).unsignCookie(raw);
  return unsigned.valid ? unsigned.value : null;
}

File: biglaw-go/internal/auth/oauth.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2026 Discover Legal

// OAuth 2.0 / OpenID Connect login flows for Google, Microsoft (common
// tenant), and LinkedIn. Authorization-code flow with no extra OAuth SDK:
// build the authorize URL, exchange the code for tokens, fetch userinfo,
// normalise the identity. Configured purely from cfg.Auth (env-derived);
// the HTTP handlers live in internal/api/auth.go.

package auth

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/discover-legal/biglaw-go/internal/config"
)

// Provider names an OAuth/OIDC login provider. The string value doubles as
// the URL path segment (/auth/<provider>/login, /auth/<provider>/callback).
type Provider string

const (
	ProviderGoogle    Provider = "google"
	ProviderMicrosoft Provider = "microsoft"
	ProviderLinkedIn  Provider = "linkedin"
)

// Providers lists every supported provider in display order.
var Providers = []Provider{ProviderGoogle, ProviderMicrosoft, ProviderLinkedIn}

// Identity is the normalised result of a provider userinfo call.
type Identity struct {
	Sub           string
	Email         string
	Name          string
	Picture       string // avatar URL; empty when the provider omits it
	EmailVerified bool
}

type providerSpec struct {
	authURL     string
	tokenURL    string
	userInfoURL string
	scope       string
}

var providerSpecs = map[Provider]providerSpec{
	ProviderGoogle: {
		authURL:     "https://accounts.google.com/o/oauth2/v2/auth",
		tokenURL:    "https://oauth2.googleapis.com/token",
		userInfoURL: "https://openidconnect.googleapis.com/v1/userinfo",
		scope:       "openid email profile",
	},
	ProviderMicrosoft: {
		authURL:     "https://login.microsoftonline.com/common/oauth2/v2.0/authorize",
		tokenURL:    "https://login.microsoftonline.com/common/oauth2/v2.0/token",
		userInfoURL: "https://graph.microsoft.com/oidc/userinfo",
		scope:       "openid email profile",
	},
	ProviderLinkedIn: {
		authURL:     "https://www.linkedin.com/oauth/v2/authorization",
		tokenURL:    "https://www.linkedin.com/oauth/v2/accessToken",
		userInfoURL: "https://api.linkedin.com/v2/userinfo",
		scope:       "openid email profile",
	},
}

// oauthHTTP is the shared client for token + userinfo calls. 30 s mirrors the
// connector timeout used elsewhere.
var oauthHTTP = &http.Client{Timeout: 30 * time.Second}

// Credentials returns the configured client ID + secret for a provider.
func Credentials(cfg *config.Config, p Provider) (clientID, clientSecret string) {
	switch p {
	case ProviderGoogle:
		return cfg.Auth.GoogleClientID, cfg.Auth.GoogleClientSecret
	case ProviderMicrosoft:
		return cfg.Auth.MicrosoftClientID, cfg.Auth.MicrosoftClientSecret
	case ProviderLinkedIn:
		return cfg.Auth.LinkedInClientID, cfg.Auth.LinkedInClientSecret
	}
	return "", ""
}

// IsConfigured reports whether both client ID and secret are set for p.
func IsConfigured(cfg *config.Config, p Provider) bool {
	id, secret := Credentials(cfg, p)
	return id != "" && secret != ""
}

// ConfiguredProviders returns the names of every provider with credentials —
// the UI uses this to decide which login buttons to render.
func ConfiguredProviders(cfg *config.Config) []string {
	out := []string{}
	for _, p := range Providers {
		if IsConfigured(cfg, p) {
			out = append(out, string(p))
		}
	}
	return out
}

// RedirectURI is the callback URL registered with the provider:
// {PUBLIC_BASE_URL}/auth/{provider}/callback.
func RedirectURI(cfg *config.Config, p Provider) string {
	return strings.TrimRight(cfg.Auth.BaseURL, "/") + "/auth/" + string(p) + "/callback"
}

// AuthorizeURL builds the provider authorize URL for the login redirect,
// carrying the anti-CSRF state and the OIDC scopes.
func AuthorizeURL(cfg *config.Config, p Provider, state string) (string, error) {
	spec, ok := providerSpecs[p]
	if !ok {
		return "", fmt.Errorf("unknown OAuth provider %q", p)
	}
	clientID, _ := Credentials(cfg, p)
	u, err := url.Parse(spec.authURL)
	if err != nil {
		return "", err
	}
	q := u.Query()
	q.Set("response_type", "code")
	q.Set("client_id", clientID)
	q.Set("redirect_uri", RedirectURI(cfg, p))
	q.Set("scope", spec.scope)
	q.Set("state", state)
	u.RawQuery = q.Encode()
	return u.String(), nil
}

// ExchangeCode swaps the authorization code for an access token.
// Error messages never include credentials or response bodies.
func ExchangeCode(ctx context.Context, cfg *config.Config, p Provider, code string) (string, error) {
	spec, ok := providerSpecs[p]
	if !ok {
		return "", fmt.Errorf("unknown OAuth provider %q", p)
	}
	clientID, clientSecret := Credentials(cfg, p)
	form := url.Values{
		"grant_type":    {"authorization_code"},
		"code":          {code},
		"redirect_uri":  {RedirectURI(cfg, p)},
		"client_id":     {clientID},
		"client_secret": {clientSecret},
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, spec.tokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	res, err := oauthHTTP.Do(req)
	if err != nil {
		return "", fmt.Errorf("token exchange request failed: %w", err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return "", fmt.Errorf("token exchange failed (%d)", res.StatusCode)
	}
	var token struct {
		AccessToken string `json:"access_token"`
	}
	if err := json.NewDecoder(io.LimitReader(res.Body, 1<<20)).Decode(&token); err != nil {
		return "", fmt.Errorf("token exchange returned invalid JSON")
	}
	if token.AccessToken == "" {
		return "", fmt.Errorf("no access_token in response")
	}
	return token.AccessToken, nil
}

// FetchUserInfo calls the provider's OIDC userinfo endpoint and maps the
// payload to a normalised Identity. Microsoft sometimes omits "email" and
// returns the address as "preferred_username" instead.
func FetchUserInfo(ctx context.Context, p Provider, accessToken string) (*Identity, error) {
	spec, ok := providerSpecs[p]
	if !ok {
		return nil, fmt.Errorf("unknown OAuth provider %q", p)
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, spec.userInfoURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)

	res, err := oauthHTTP.Do(req)
	if err != nil {
		return nil, fmt.Errorf("userinfo request failed: %w", err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("userinfo failed (%d)", res.StatusCode)
	}
	var info map[string]interface{}
	if err := json.NewDecoder(io.LimitReader(res.Body, 1<<20)).Decode(&info); err != nil {
		return nil, fmt.Errorf("userinfo returned invalid JSON")
	}

	email := jsonStr(info, "email")
	if email == "" && p == ProviderMicrosoft {
		email = jsonStr(info, "preferred_username")
	}
	name := jsonStr(info, "name")
	if name == "" {
		name = email
	}
	if name == "" {
		name = "User"
	}
	return &Identity{
		Sub:     jsonStr(info, "sub"),
		Email:   email,
		Name:    name,
		Picture: jsonStr(info, "picture"),
		// Treat absent email_verified as verified (LinkedIn omits it);
		// only an explicit false rejects the login.
		EmailVerified: info["email_verified"] != false,
	}, nil
}

func jsonStr(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}
